src/import-data/useImportRequest.ts

```diff
@@ -83,50 +85,75 @@ const getCatalogDatasetImportRequest = (queryParams: QueryParams): CatalogDatase
+  const snapshots = catalogDatasets
+    .filter((dataset) => snapshotIds.includes(dataset['dct:identifier']))
+    .map((dataset) => {
+      return {
+        // The previous step filters the list to only datasets with 'dct:identifier' defined
+        id: dataset['dct:identifier']!,
+        title: dataset['dct:title'],
+        description: dataset['dct:description'],
+      };
+    });
+
   return {
     type: 'catalog-snapshots',
-    snapshotIds,
+    snapshots,
   };
 };
 
-export const getImportRequest = (queryParams: QueryParams): ImportRequest => {
+export const getImportRequest = (queryParams: QueryParams): Promise<ImportRequest> => {
   const format = getFormat(queryParams);
 
   switch (format) {
     case 'pfb':
-      return getFileImportRequest(queryParams, 'pfb');
+      return Promise.resolve(getFileImportRequest(queryParams, 'pfb'));
     case 'bagit':
-      return getFileImportRequest(queryParams, 'bagit');
+      return Promise.resolve(getFileImportRequest(queryParams, 'bagit'));
     case 'entitiesjson':
-      return getFileImportRequest(queryParams, 'entities');
+      return Promise.resolve(getFileImportRequest(queryParams, 'entities'));
     case 'tdrexport':
-      return getTDRSnapshotExportImportRequest(queryParams);
+      return Promise.resolve(getTDRSnapshotExportImportRequest(queryParams));
     case 'snapshot':
       if (queryParams.snapshotIds) {
         return getCatalogSnapshotsImportRequest(queryParams);
       }
-      return getTDRSnapshotReferenceImportRequest(queryParams);
+      return Promise.resolve(getTDRSnapshotReferenceImportRequest(queryParams));
     case 'catalog':
-      return getCatalogDatasetImportRequest(queryParams);
+      return Promise.resolve(getCatalogDatasetImportRequest(queryParams));
     default:
       throw new Error(`Invalid format: ${format}`);
   }
 };
 
-export type UseImportRequestResult = { isValid: true; importRequest: ImportRequest } | { isValid: false; error: Error };
+export type UseImportRequestResult =
+  | { status: 'Loading' }
+  | { status: 'Ready'; importRequest: ImportRequest }
+  | { status: 'Error'; error: Error };
 
 export const useImportRequest = (): UseImportRequestResult => {
   const { query } = useRoute();
-  try {
-    const importRequest = getImportRequest(query);
-    return { isValid: true, importRequest };
-  } catch (originalError: unknown) {
-    const error = originalError instanceof Error ? originalError : new Error('Unknown error');
-    return { isValid: false, error };
-  }
+
+  const [result, setResult] = useState<UseImportRequestResult>({ status: 'Loading' });
+  useEffect(() => {
+    (async () => {
+      try {
+        setResult({ status: 'Loading' });
+        const importRequest = await getImportRequest(query);
+        setResult({ status: 'Ready', importRequest });
+      } catch (originalError: unknown) {
+        const error = originalError instanceof Error ? originalError : new Error('Unknown error');
```

Comment:
Feels bad to lose whatever error information is provided in `unknown`, can you do anything reasonable with it to make it available/visible for troubleshooting purposes?

Reply:
Because JS allows throwing any type of object, TypeScript requires the caught object to be typed either `any` or `unknown`. My thinking here was that `getImportRequest` is expected to only throw `Error`s (in case of an invalid request), so this was as much to appease TS as anything.
https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript

However, now that you mention it, I think the `fetchDataCatalog` function may throw a `Response`, so that needs to be caught and re-thrown as an error.

Reply:
Added more error handling in https://github.com/DataBiosphere/terra-ui/pull/4307/commits/3d2eca96342d6bb2f201d7db98b568c480c09e17.